📚 Translate calendar related strings

lelemm opened the pull request:
Translate calendar related strings

joel-jeremy commented on packages/desktop-client/src/hooks/useSyncLocale.ts:
```diff
@@ -0,0 +1,15 @@
+import { useEffect } from 'react';
+import { useDispatch } from 'react-redux';
```
This should import the `useDispatch` from `src/redux` instead which is properly typed with the correct state and actions.

This is one main source of confusion that I wanted to avoid in https://github.com/actualbudget/actual/pull/4101. I wonder why the eslint didn't report an error/warning when you imported the untyped useDispatch... There is a rule setup for that.

lelemm replied:
Thanks!

lelemm commented:
@joel-jeremy , any way to access the redux store from loot-core?
There are some dates that would need translation there:

https://github.com/actualbudget/actual/blob/1442662eb7ba8901835e0cb173f05430b7edcaa6/packages/loot-core/src/server/budget/actions.ts#L554-L556

and
https://github.com/actualbudget/actual/blob/1442662eb7ba8901835e0cb173f05430b7edcaa6/packages/loot-core/src/shared/schedules.ts#L62-L194

joel-jeremy commented:
> @joel-jeremy , any way to access the redux store from loot-core? There are some dates that would need translation there:
> 
> https://github.com/actualbudget/actual/blob/1442662eb7ba8901835e0cb173f05430b7edcaa6/packages/loot-core/src/server/budget/actions.ts#L554-L556
> 
> and
> 
> https://github.com/actualbudget/actual/blob/1442662eb7ba8901835e0cb173f05430b7edcaa6/packages/loot-core/src/shared/schedules.ts#L62-L194
> 
> There both cases of translations AND dates needing locale (from the redux store)

You can actually import the store the call `store.getState()` to get the data you need but this is tricky because we don't want the loot-core server to have a dependency on the client store, if possible IMO.

Why do we need to put the locale in the store? Why don't we store the locale as a global pref (same as how the language global pref is being stored) and get the locale from there?

lelemm commented:
> > @joel-jeremy , any way to access the redux store from loot-core? There are some dates that would need translation there:
> > https://github.com/actualbudget/actual/blob/1442662eb7ba8901835e0cb173f05430b7edcaa6/packages/loot-core/src/server/budget/actions.ts#L554-L556
> > 
> > and
> > https://github.com/actualbudget/actual/blob/1442662eb7ba8901835e0cb173f05430b7edcaa6/packages/loot-core/src/shared/schedules.ts#L62-L194
> > 
> > There both cases of translations AND dates needing locale (from the redux store)
> 
> You can actually import the store the call `store.getState()` to get the data you need but this is tricky because we don't want the loot-core server to have a dependency on the client store, if possible IMO.
> 
> Why do we need to put the locale in the store? Why don't we store the locale as a global pref (same as how the language global pref is being stored) and get the locale from there?

I guess I can try serializing the locale object to see what happens. 
Also, can you link me an example of how can I access the global pref from lootcore?

joel-jeremy commented:
> > > @joel-jeremy , any way to access the redux store from loot-core? There are some dates that would need translation there:
> > > https://github.com/actualbudget/actual/blob/1442662eb7ba8901835e0cb173f05430b7edcaa6/packages/loot-core/src/server/budget/actions.ts#L554-L556
> > > 
> > > and
> > > https://github.com/actualbudget/actual/blob/1442662eb7ba8901835e0cb173f05430b7edcaa6/packages/loot-core/src/shared/schedules.ts#L62-L194
> > > 
> > > There both cases of translations AND dates needing locale (from the redux store)
> > 
> > 
> > You can actually import the store the call `store.getState()` to get the data you need but this is tricky because we don't want the loot-core server to have a dependency on the client store, if possible IMO.
> > Why do we need to put the locale in the store? Why don't we store the locale as a global pref (same as how the language global pref is being stored) and get the locale from there?
> 
> I guess I can try serializing the locale object to see what happens. Also, can you link me an example of how can I access the global pref from lootcore?

Global prefs are store via `asyncStorage` so I believe you can import `platform/server/asyncStorage` and call `asyncStorage.getItem`

lelemm commented:
> actual/packages/loot-core/src/shared/schedules.ts

tried to use asyncStorage with `date-fns/locale/enUS`, didn't work, because there are methods on it. So I kept the redux store for client code (to minimize the number fo calls to get the locale from the module) and get the locale for the language for every rule description on the loot-core at runtime

lelemm commented:
Also, anyone knows why
api and electron is failing to build? when I tried to build locally gave me the same error "ERROR in Conflict: Multiple assets emit different content to the same filename bundle.api.js.map" and then it created this file on the `api/app/`:  `vendors-node_modules_date-fns_esm_locale_index_js.bundle.api.js`

joel-jeremy commented:
> @joel-jeremy, I tried to use asyncStorage with `date-fns/locale/enUS`, didn't work, because there are methods on it. So I kept the redux store for client code (to minimize the number fo calls to get the locale from the module) and get the locale from the language for every rule description on the loot-core at runtime

We shouldn't be storing the entire Locale object IMO. Instead, we should store the locale key on the global pref and retrieve the Locale instance based on that key. 

It is also not recommended to store Locale in the store if it contains non-serializable properties such as functions: https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions

lelemm commented:
> > @joel-jeremy, I tried to use asyncStorage with `date-fns/locale/enUS`, didn't work, because there are methods on it. So I kept the redux store for client code (to minimize the number fo calls to get the locale from the module) and get the locale from the language for every rule description on the loot-core at runtime
> 
> We shouldn't be storing the entire Locale object IMO. Instead, we should store the locale key on the global pref and retrieve the Locale instance based on that key.
> 
> It is also not recommended to store Locale in the store if it contains non-serializable properties such as functions: https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions

I removed all reference to redux state related to locale, also removed the dynamic locale loading, now its a switch/case (it was preventing api to compile)

xthiago reviewed (commented):
@lelemm I tested your branch and I discovered one issue 🐛 related to default language recognition (HTTP Request header `Accept-Language`). I am not sure if the previous reviews already covered it.

When I have the following configuration (default):

- Language: System default // which is default.

![brave_GCaSu3PFuV](https://github.com/user-attachments/assets/44e99105-72fe-4880-8371-c6be918875b7)

And I access the Actual with `Accept-Language: pt-BR`, I got:

- The texts are translated to my preferred language, but the dates not.

![brave_cymsU15Cea](https://github.com/user-attachments/assets/cea43866-7d9f-4e7d-b3c4-b0ab0c5994fe)
![brave_go0MWSpQUs](https://github.com/user-attachments/assets/5d2d0bd0-6e91-4ff8-b21b-5f0fa31dc5a0)

Expected result: Everything (including the dates/months) must be translated to the default language (if a translation for it exists).

PS: If you want a way to quickly change the preferred language, you can use this: https://github.com/locale-switcher/locale-switcher/

xthiago commented:
@lelemm yes, it works. Thanks for the fix. 🎉

Actually, I located another issue not directly related to your PR (also happens with the regular texts), so I can open a separated issue for it.

Browsers offer user's the ability to specify more than one preferable language and allow a prioritization. Like this one:

![image](https://github.com/user-attachments/assets/65db3ad7-5686-4443-b484-562372d03bf1)

I did one test setting: 1. Netherlands, Spanish, Brazilian Portuguese and then US English:

```
accept-language: nl-NL, es;q=0.9, pt-BR;q=0.8, en-US;q=0.7
```

**Expected result:** Actual should check the available translations accordingly with the priority list and use it.

**Actual result:** If more than one preferable language is set, the UI ignores it and always sets to English (no matter if there is a translation for the chosen languages or not).

lelemm commented:
Judging by the time ran to try to do E2E, this might broke playwright. Any ideas?

matt-fidd commented:
/update-vrt

matt-fidd commented:
It's running now, but there's a test or two to fix

lelemm commented:
> It's running now, but there's a test or two to fix

ty, I will check it

lelemm commented:
@matt-fidd where are stored the reference images? I don't remember

matt-fidd commented:
> @matt-fidd where are stored the reference images? I don't remember

packages/desktop-client/e2e

lelemm commented:
![Mobile-Budget-Envelope-applies-budget-template-1-chromium-linux](https://github.com/user-attachments/assets/d460d900-a313-4b8e-b891-5d0d40e53345)
![test-failed-1](https://github.com/user-attachments/assets/2722d7b6-f4df-4270-b189-24a51bc2eb81)
no diffs, wtf

matt-fidd commented:
Looks like it's not happy about there being two alerts, not sure why that didn't fail on an earlier PR because it's not a change in here that's done it

<img width="757" alt="image" src="https://github.com/user-attachments/assets/d72afab1-dda5-4525-92f3-99263f5afa4e" />

matt-fidd commented:
https://github.com/actualbudget/actual/pull/4223

matt-fidd commented:
> **Expected result:** Actual should check the available translations accordingly with the priority list and use it.
> 
> **Actual result:** If more than one preferable language is set, the UI ignores it and always sets to English (no matter if there is a translation for the chosen languages or not).

@xthiago  Please could you open a separate feature request for that?

matt-fidd commented:
@lelemm sorted :)

lelemm commented:
ty @matt-fidd

youngcw commented:
You can split it if you want.  I just wanted to get eyes on it

lelemm commented:
I'm not splitting, This is ready for review

matt-fidd commented on packages/loot-core/src/shared/months.ts:
```diff
@@ -393,12 +393,16 @@ export function sheetForMonth(month: string): string {
 
-export function format(month: DateLike, format: string): string {
-  return d.format(_parse(month), format);
+export function format(
+  month: DateLike,
+  format: string,
+  locale?: Locale,
+): string {
+  return d.format(_parse(month), format, { locale });
```
Is there a way we can grab the locale in here instead? I'm really not a fan of an additional parameter all over the codebase if we can avoid it.

What about a loot-core helper that uses asyncStorage (platform/server/asyncStorage) instead of making desktop-client do the work?

```js
asyncStorage.getItem('language')
```

lelemm replied:
I guess we can. The main reason to mark all functions is to explicitly use locales when really needed.

matt-fidd replied:
For others, we had a go at this but it would mean making the format function async - which isn't really any better than updating the function calls everywhere anyway, and awaiting a formatting function feels strange.

lelemm commented:
/update-vrt

MatissJanis commented on packages/desktop-client/src/components/settings/LanguageSettings.tsx:
```diff
@@ -12,16 +12,22 @@ import { Select, type SelectOption } from '../common/Select';
 
 import { Setting } from './UI';
 
+const languageDisplayNameOverride: { [key: string]: string } = {
+  'pt-BR': 'Português (Brasil)',
+};
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏why do we need this?

lelemm replied:
This is a way so we can override the browser default string (not capitalized) and to not force a capitalized for non roman languages, we talked about leaving a possible override array.
It was discussed here https://discord.com/channels/937901803608096828/1266785850025312379/1330892905836970067 to here https://discord.com/channels/937901803608096828/1266785850025312379/1331652100253417543
